src/ArmTemplates/Extractor/EntityExtractors/ServiceApisProductsExtractor.cs

```diff
@@ -0,0 +1,136 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.API.Clients.Abstractions;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Constants;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Exceptions;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Extensions;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates.Abstractions;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Extractor.EntityExtractors.Abstractions;
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Extractor.Models;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Extractor.EntityExtractors
+{
+    public class ServiceApisProductsExtractor : TemplateGeneratorBase, IServiceApisProductsExtractor
+    {
+        readonly ILogger<ServiceApisProductsExtractor> logger;
+        
+        readonly IServiceApiProductsApiClient serviceApiProductsApiClient;
+        readonly IServiceApisApiClient serviceApisApiClient;
+
+        public ServiceApisProductsExtractor(
+            ILogger<ServiceApisProductsExtractor> logger,
+            IServiceApiProductsApiClient productApisApiClient,
+            IServiceApisApiClient serviceApiClient)
+        {
+            this.logger = logger;
+            
+            this.serviceApiProductsApiClient = productApisApiClient;
+            this.serviceApisApiClient = serviceApiClient;
+        }
+
+        public async Task<Template> GenerateServiceApisProductsARMTemplateAsync(string singleApiName, List<string> multipleApiNames, ExtractorParameters extractorParameters)
+        {
+            Template armTemplate = this.GenerateEmptyPropertyTemplateWithParameters();
+            List<TemplateResource> templateResources = new List<TemplateResource>();
+
+            if (!string.IsNullOrEmpty(singleApiName))
+            {
+                try
+                {
+                    var serviceApi = await this.serviceApisApiClient.GetSingleServiceApiAsync(
+                        extractorParameters.SourceApimName, 
+                        extractorParameters.ResourceGroup, 
+                        singleApiName);
+
+                    if (serviceApi is null)
+                    {
+                        throw new ServiceApiNotFoundException($"ServiceApi with name '{singleApiName}' not found");
+                    }
+
+                    this.logger.LogInformation("{0} Product API found ...", singleApiName);
+                    templateResources.AddRange(await this.GenerateServiceApiProductsTemplateResourcesAsync(singleApiName, extractorParameters, new string[] { }));
```

Comment:
I prefer not to do inline method calls, definately not if they are async

Reply:
same for me, thanks for noticing